In [1]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns
%matplotlib inline

In [2]:
data=pd.read_csv("https://raw.githubusercontent.com/aniruddhachoudhury/Red-Wine-Quality/master/winequality-red.csv")

In [3]:
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [4]:
data['quality'].unique()

array([5, 6, 7, 4, 8, 3])

In [5]:
data['quality'].value_counts()

5    681
6    638
7    199
4     53
8     18
3     10
Name: quality, dtype: int64

In [6]:
data.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [8]:
data.describe().T

count       mean        std      min      25%  \
fixed acidity         1599.0   8.319637   1.741096  4.60000   7.1000   
volatile acidity      1599.0   0.527821   0.179060  0.12000   0.3900   
citric acid           1599.0   0.270976   0.194801  0.00000   0.0900   
residual sugar        1599.0   2.538806   1.409928  0.90000   1.9000   
chlorides             1599.0   0.087467   0.047065  0.01200   0.0700   
free sulfur dioxide   1599.0  15.874922  10.460157  1.00000   7.0000   
total sulfur dioxide  1599.0  46.467792  32.895324  6.00000  22.0000   
density               1599.0   0.996747   0.001887  0.99007   0.9956   
pH                    1599.0   3.311113   0.154386  2.74000   3.2100   
sulphates             1599.0   0.658149   0.169507  0.33000   0.5500   
alcohol               1599.0  10.422983   1.065668  8.40000   9.5000   
quality               1599.0   5.636023   0.807569  3.00000   5.0000   

                           50%        75%        max  
fixed acidity          7.90000   9.200000   15.90000  
volatile acidity       0.52000   0.640000    1.58000  
citric acid            0.26000   0.420000    1.00000  
residual sugar         2.20000   2.600000   15.50000  
chlorides              0.07900   0.090000    0.61100  
free sulfur dioxide   14.00000  21.000000   72.00000  
total sulfur dioxide  38.00000  62.000000  289.00000  
density                0.99675   0.997835    1.00369  
pH                     3.31000   3.400000    4.01000  
sulphates              0.62000   0.730000    2.00000  
alcohol               10.20000  11.100000   14.90000  
quality                6.00000   6.000000    8.00000

In [9]:
X=data.drop("quality",axis=1)

In [10]:
X.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  
0      9.4  
1      9.8  
2      9.8  
3      9.8  
4      9.4

In [11]:
y = data['quality']
y.head()

0    5
1    5
2    5
3    6
4    5
Name: quality, dtype: int64

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [13]:
X_train.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
548            12.4             0.350         0.49             2.6      0.079   
355             6.7             0.750         0.01             2.4      0.078   
1296            6.6             0.630         0.00             4.3      0.093   
209            11.0             0.300         0.58             2.1      0.054   
140             8.4             0.745         0.11             1.9      0.090   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
548                  27.0                  69.0  0.99940  3.12       0.75   
355                  17.0                  32.0  0.99550  3.55       0.61   
1296                 51.0                  77.5  0.99558  3.20       0.45   
209                   7.0                  19.0  0.99800  3.31       0.88   
140                  16.0                  63.0  0.99650  3.19       0.82   

      alcohol  
548      10.4  
355      12.8  
1296      9.5  
209      10.5  
140       9.6

In [14]:
y_train.head()

548     6
355     6
1296    5
209     7
140     5
Name: quality, dtype: int64

In [15]:
from sklearn.preprocessing import StandardScaler

In [16]:
scaler = StandardScaler()

In [17]:
scaler.fit(X_train)

StandardScaler()

In [18]:
scaler.mean_

array([ 8.30345472,  0.53246499,  0.26933707,  2.54691877,  0.08772736,
       15.91223156, 46.76330532,  0.99677933,  3.31453782,  0.65881419,
       10.41521942])

In [19]:
X_train_tf=scaler.transform(X_train)

In [20]:
from sklearn.svm import SVC
model=SVC()

In [21]:
model.fit(X_train_tf,y_train)

SVC()

In [22]:
model.score(X_train_tf,y_train)

0.6778711484593838

In [23]:
X_test_tf=scaler.transform(X_test)

In [24]:
y_predict=model.predict(X_test_tf)

In [25]:
y_test.head()

803     6
124     5
350     6
682     5
1326    6
Name: quality, dtype: int64

In [26]:
from sklearn.metrics import accuracy_score

In [27]:
accuracy_score(y_test,y_predict)

0.5984848484848485

In [28]:
from sklearn.linear_model import LogisticRegression

In [29]:
model2=LogisticRegression()

In [30]:
model2.fit(X_train_tf,y_train)

LogisticRegression()

In [31]:
y_predict2=model2.predict(X_test_tf)

In [32]:
accuracy_score(y_test,y_predict2)

0.571969696969697

In [33]:
X_test_tf[0]

array([-0.3536421 ,  0.15558944, -0.96737373, -0.03334372,  0.55556956,
       -0.18596079, -0.02314512,  0.1740298 , -0.48314224,  0.00685666,
       -0.76696884])

In [34]:
model.predict([[-0.3536421 ,  0.15558944, -0.96737373, -0.03334372,  0.55556956,
       -0.18596079, -0.02314512,  0.1740298 , -0.48314224,  0.00685666,
       -0.76696884]])

array([5])

In [35]:
#gridsearch CV
from sklearn.model_selection import GridSearchCV

In [45]:
# defining parameter range
param_grid = {'C': [ 1, 10, 100], 
              'gamma': [1, 0.1],
              'kernel': ['rbf','linear']}

In [46]:
# GridSearchCV(estimator,paramgrid,)
grid = GridSearchCV(SVC(), param_grid, refit = True, verbose = 3)

In [47]:
grid.fit(X_train,y_train)

Fitting 5 folds for each of 12 candidates, totalling 60 fits
[CV 1/5] END ..........C=1, gamma=1, kernel=rbf;, score=0.535 total time=   0.2s
[CV 2/5] END ..........C=1, gamma=1, kernel=rbf;, score=0.542 total time=   0.2s
[CV 3/5] END ..........C=1, gamma=1, kernel=rbf;, score=0.542 total time=   0.2s
[CV 4/5] END ..........C=1, gamma=1, kernel=rbf;, score=0.589 total time=   0.2s
[CV 5/5] END ..........C=1, gamma=1, kernel=rbf;, score=0.542 total time=   0.2s
[CV 1/5] END .......C=1, gamma=1, kernel=linear;, score=0.530 total time=   5.4s
[CV 2/5] END .......C=1, gamma=1, kernel=linear;, score=0.579 total time=   8.4s
[CV 3/5] END .......C=1, gamma=1, kernel=linear;, score=0.551 total time=  10.0s
[CV 4/5] END .......C=1, gamma=1, kernel=linear;, score=0.612 total time=   7.5s
[CV 5/5] END .......C=1, gamma=1, kernel=linear;, score=0.650 total time=  10.5s
[CV 1/5] END ........C=1, gamma=0.1, kernel=rbf;, score=0.544 total time=   0.2s
[CV 2/5] END ........C=1, gamma=0.1, kernel=rbf;

GridSearchCV(estimator=SVC(),
             param_grid={'C': [1, 10, 100], 'gamma': [1, 0.1],
                         'kernel': ['rbf', 'linear']},
             verbose=3)

In [48]:
print(grid.best_params_)

{'C': 100, 'gamma': 1, 'kernel': 'linear'}


In [49]:
grid.best_estimator_

SVC(C=100, gamma=1, kernel='linear')

In [50]:
grid_predict = grid.predict(X_test)

In [51]:
grid_predict

array([5, 5, 6, 5, 6, 5, 5, 5, 6, 6, 6, 5, 5, 5, 5, 6, 5, 5, 6, 5, 5, 5,
       6, 6, 5, 5, 6, 5, 5, 6, 5, 5, 6, 5, 6, 5, 6, 6, 5, 6, 5, 5, 6, 5,
       6, 6, 6, 5, 5, 6, 5, 5, 6, 6, 5, 5, 6, 5, 6, 5, 5, 6, 5, 5, 6, 5,
       6, 5, 6, 5, 6, 5, 6, 6, 6, 5, 6, 6, 6, 6, 5, 6, 5, 6, 6, 6, 5, 6,
       6, 5, 6, 5, 6, 6, 5, 5, 5, 6, 5, 6, 5, 5, 6, 6, 6, 6, 5, 5, 6, 5,
       6, 5, 6, 5, 6, 6, 6, 6, 5, 6, 6, 5, 6, 5, 5, 5, 6, 6, 6, 6, 6, 5,
       5, 6, 6, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 5, 6, 5, 6, 5, 6, 6, 5, 6,
       6, 6, 5, 6, 5, 6, 6, 6, 6, 5, 5, 6, 5, 5, 5, 5, 5, 5, 6, 5, 5, 6,
       6, 5, 5, 5, 5, 6, 5, 6, 5, 6, 6, 6, 6, 5, 6, 6, 6, 6, 6, 5, 5, 5,
       5, 6, 5, 5, 5, 5, 6, 6, 5, 5, 5, 6, 6, 5, 6, 6, 6, 6, 5, 5, 6, 5,
       5, 6, 6, 6, 5, 5, 5, 6, 5, 5, 5, 5, 6, 6, 6, 6, 5, 6, 5, 5, 5, 5,
       6, 6, 5, 5, 6, 5, 6, 5, 6, 6, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 5, 6,
       6, 6, 5, 5, 6, 6, 5, 6, 5, 5, 5, 5, 6, 6, 6, 5, 6, 5, 5, 5, 5, 6,
       5, 6, 5, 6, 5, 6, 5, 5, 5, 6, 5, 6, 6, 6, 5,

In [52]:
accuracy_score(y_test,grid_predict)

0.5643939393939394

**After hyperparametertuening best patameters {'C': 100, 'gamma': 1, 'kernel': 'linear'} and accuracy score: 0.5643939393939394**

In [53]:
from sklearn.metrics import classification_report,confusion_matrix

In [55]:
print(classification_report(y_test,y_predict))

              precision    recall  f1-score   support

           3       0.00      0.00      0.00         2
           4       0.00      0.00      0.00        19
           5       0.65      0.78      0.71       217
           6       0.55      0.60      0.57       213
           7       0.56      0.27      0.37        70
           8       0.00      0.00      0.00         7

    accuracy                           0.60       528
   macro avg       0.29      0.28      0.27       528
weighted avg       0.56      0.60      0.57       528



/opt/conda/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [56]:
print(classification_report(y_test,grid_predict))

              precision    recall  f1-score   support

           3       0.00      0.00      0.00         2
           4       0.00      0.00      0.00        19
           5       0.62      0.76      0.69       217
           6       0.50      0.62      0.56       213
           7       0.00      0.00      0.00        70
           8       0.00      0.00      0.00         7

    accuracy                           0.56       528
   macro avg       0.19      0.23      0.21       528
weighted avg       0.46      0.56      0.51       528



/opt/conda/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1334: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
